Allow RPC to be initialized again after shutdown.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42723 Allow RPC to be initialized again after shutdown.**

This PR is addressing https://github.com/pytorch/pytorch/issues/39340
and allows users to initialize RPC again after shutdown. Major changes in the
PR include:

1. Change to DistAutogradContainer to support this.
2. Ensure PythonRpcHandler is reinitialized appropriately.
3. Use PrefixStore in RPC initialization to ensure each new `init_rpc` uses a
different prefix.

Differential Revision: [D22993909](https://our.internmc.facebook.com/intern/diff/D22993909/)

pritamdamania committed:
Allow RPC to be initialized again after shutdown.

This PR is addressing https://github.com/pytorch/pytorch/issues/39340
and allows users to initialize RPC again after shutdown. Major changes in the
PR include:

1. Change to DistAutogradContainer to support this.
2. Ensure PythonRpcHandler is reinitialized appropriately.
3. Use PrefixStore in RPC initialization to ensure each new `init_rpc` uses a
different prefix.

Differential Revision: [D22993909](https://our.internmc.facebook.com/intern/diff/D22993909/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/autograd/context/container.cpp:
```diff
@@ -41,8 +41,16 @@ DistAutogradContainer& DistAutogradContainer::init(int64_t worker_id) {
   TORCH_CHECK(
-      !container.initialized_,
-      "Container is already initialized! Cannot initialize it twice!");
+      !container.initialized_ || (worker_id == container.worker_id_),
```
Does this mean if we shut down RPC, and then re-init, we can't change the passed in rank? I don't think there's a reason to do so, but just confirming.

pritamdamania87 replied:
Yes, that is correct. I think it is probably safer to not allow rank changes and its easier to also just reuse the same container.

rohan-varma commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -68,27 +68,33 @@ void cleanupPyObj(py::object& obj) {
-  typeParser_ = std::make_shared<jit::ScriptTypeParser>(
-      std::make_shared<PythonTypeResolver>());
+void PythonRpcHandler::init() {
+  if (!initialized_) {
```
I'm wondering what's the motivation for this atomic and if it's to prevent races? If there could be thread races, this atomic won't be enough to ensure initialization only once right? If there's no races, why do we need an atomic?

pritamdamania87 replied:
This function can probably be executed twice and that is safe and would happen rarely. Apart from that, the atomic is mostly to act as a memory barrier to ensure other threads are informed about this change. For example, if one thread calls `cleanup` and another thread calls `getInstance` after it, if we don't use an atomic the thread calling `getInstance` might still see the old value of `initialized`.

rohan-varma reviewed (commented):
Looks good, had a couple of questions and there are some test failures.

pritamdamania committed:
Update on "Allow RPC to be initialized again after shutdown."


This PR is addressing https://github.com/pytorch/pytorch/issues/39340
and allows users to initialize RPC again after shutdown. Major changes in the
PR include:

1. Change to DistAutogradContainer to support this.
2. Ensure PythonRpcHandler is reinitialized appropriately.
3. Use PrefixStore in RPC initialization to ensure each new `init_rpc` uses a
different prefix.

Differential Revision: [D22993909](https://our.internmc.facebook.com/intern/diff/D22993909/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM!

pritamdamania committed:
Update on "Allow RPC to be initialized again after shutdown."


This PR is addressing https://github.com/pytorch/pytorch/issues/39340
and allows users to initialize RPC again after shutdown. Major changes in the
PR include:

1. Change to DistAutogradContainer to support this.
2. Ensure PythonRpcHandler is reinitialized appropriately.
3. Use PrefixStore in RPC initialization to ensure each new `init_rpc` uses a
different prefix.

Differential Revision: [D22993909](https://our.internmc.facebook.com/intern/diff/D22993909/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -3121,6 +3121,38 @@ def test_wait_all_with_partial_exception(self):
 
+    @dist_init(setup_rpc=False)
+    def test_init_rpc_twice(self):
+        initialize_pg(self.init_method, self.rank, self.world_size)
```
It looks like we missed testing this, though it looks like this test fails depending on the init_method. For example if users want to init RPC twice with TCPStore, it currently does not work (see details in https://github.com/pytorch/pytorch/issues/46491).
